As suggested, dropping the return type from the `operator+=` etc. only when the return type is of a reference type seems to make [the above assertion failures](https://github.com/swiftlang/swift/pull/76324#issuecomment-2420145075) go away. I also tried to limit this to the this pointer case but it seems to cause quite a few other test failures maybe because the conformance checking relies on the return type being void (the details at the bottom). I think it's sufficient to drop the return type only when the return type is a reference type because that's probably what the original intention of the return-type-dropping code and a reference-type return type means it's unlikely an indirect return, which solves this `CGFunctionInfo`/`operator+=` issue with the above assert failures. Please take another look @rjmccall 

The test failures (with restricting to this pointer)
```
******************** TEST 'Swift(windows-x86_64) :: Interop/Cxx/class/protocol-conformance-typechecker.swift' FAILED ********************
S:\SourceCache\swift\test\Interop\Cxx\class\protocol-conformance-typechecker.swift:48:15: error: unexpected note produced: protocol requires function '+=' with type '(inout HasOperatorPlusEqual<CInt>, Int32) -> ()'
  static func +=(lhs: inout Self, x: Int32)
              ^
S:\SourceCache\swift\test\Interop\Cxx\class\protocol-conformance-typechecker.swift:51:1: error: unexpected error produced: type 'HasOperatorPlusEqual<CInt>' does not conform to protocol 'HasOperatorPlusEqualProtocol'
extension HasOperatorPlusEqualInt : HasOperatorPlusEqualProtocol {}
^
S:\SourceCache\swift\test\Interop\Cxx\class\protocol-conformance-typechecker.swift:51:1: error: unexpected note produced: add stubs for conformance
extension HasOperatorPlusEqualInt : HasOperatorPlusEqualProtocol {}
^

******************** TEST 'Swift(windows-x86_64) :: Interop/Cxx/stdlib/msvcprt-module-interface.swift' FAILED ********************
S:\SourceCache\swift\test\Interop\Cxx\stdlib\msvcprt-module-interface.swift:15:18: error: CHECK-STRING: expected string not found in input
// CHECK-STRING: struct basic_string<CChar, char_traits<CChar>, allocator<CChar>> : CxxRandomAccessCollection {
                 ^
<stdin>:23948:54: note: scanning from here
 static func to_wstring(_ _Val: Int32) -> std.wstring
                                                     ^

Input file: <stdin>
Check file: S:\SourceCache\swift\test\Interop\Cxx\stdlib\msvcprt-module-interface.swift

******************** TEST 'Swift(windows-x86_64) :: Interop/Cxx/stdlib/overlay/custom-collection-module-interface.swift' FAILED ********************
S:\SourceCache\swift\test\Interop\Cxx\stdlib\overlay\custom-collection-module-interface.swift:17:11: error: CHECK: expected string not found in input
// CHECK: struct SimpleCollectionReadOnly : CxxRandomAccessCollection {
          ^
<stdin>:167:2: note: scanning from here
}
 ^
<stdin>:168:1: note: possible intended match here
struct SimpleCollectionReadOnly : CxxConvertibleToCollection {
^

Input file: <stdin>
Check file: S:\SourceCache\swift\test\Interop\Cxx\stdlib\overlay\custom-collection-module-interface.swift

******************** TEST 'Swift(windows-x86_64) :: Interop/Cxx/stdlib/overlay/custom-collection.swift' FAILED ********************
S:\SourceCache\swift\test\Interop\Cxx\stdlib\overlay\custom-collection.swift:32:17: error: value of type 'SimpleCollectionReadOnly' has no member 'first'
30 | CxxCollectionTestSuite.test("SimpleCollectionReadOnly as Swift.Collection") {
31 |   let c = SimpleCollectionReadOnly()
32 |   expectEqual(c.first, 1)
   |                 `- error: value of type 'SimpleCollectionReadOnly' has no member 'first'
33 |   expectEqual(c.last, 5)
34 |

******************** TEST 'Swift(windows-x86_64) :: Interop/Cxx/stdlib/overlay/custom-iterator-module-interface.swift' FAILED ********************
S:\SourceCache\swift\test\Interop\Cxx\stdlib\overlay\custom-iterator-module-interface.swift:22:11: error: CHECK: expected string not found in input
// CHECK: struct ConstRACIteratorRefPlusEq : UnsafeCxxRandomAccessIterator, UnsafeCxxInputIterator {
          ^
<stdin>:70:2: note: scanning from here
}
 ^
<stdin>:72:1: note: possible intended match here
struct ConstRACIteratorRefPlusEq : UnsafeCxxInputIterator {
^

Input file: <stdin>
Check file: S:\SourceCache\swift\test\Interop\Cxx\stdlib\overlay\custom-iterator-module-interface.swift

******************** TEST 'Swift(windows-x86_64) :: Interop/Cxx/stdlib/use-std-span.swift' FAILED ********************
S:\SourceCache\swift\test\Interop\Cxx\stdlib\use-std-span.swift:185:18: error: cannot use mutating getter on immutable value: 's' is a 'let' constant
180 |   var arr: [Int32] = [1, 2, 3]
181 |   arr.withUnsafeMutableBufferPointer{ ubpointer in
182 |     let s = initSpan(ubpointer.baseAddress!, ubpointer.count)
    |     `- note: change 'let' to 'var' to make it mutable
183 |     expectEqual(s.size(), 3)
184 |     expectFalse(s.empty())
185 |     expectEqual(s[0], 1)
    |                  `- error: cannot use mutating getter on immutable value: 's' is a 'let' constant
186 |     expectEqual(s[1], 2)
187 |     expectEqual(s[2], 3)

******************** TEST 'Swift(windows-x86_64) :: Interop/Cxx/stdlib/use-std-string.swift' FAILED ********************
S:\SourceCache\swift\test\Interop\Cxx\stdlib\use-std-string.swift:372:14: error: for-in loop requires 'std.string' (aka 'std.basic_string<CChar, char_traits<CChar>, allocator<CChar>>') to conform to 'Sequence'
370 |     let cxx1 = std.string()
371 |     var iterated = false
372 |     for _ in cxx1 {
    |              `- error: for-in loop requires 'std.string' (aka 'std.basic_string<CChar, char_traits<CChar>, allocator<CChar>>') to conform to 'Sequence'
373 |         iterated = true
374 |     }

******************** TEST 'Swift(windows-x86_64) :: Interop/Cxx/stdlib/use-std-vector.swift' FAILED ********************
S:\SourceCache\swift\test\Interop\Cxx\stdlib\use-std-vector.swift:32:20: error: cannot convert value of type '[Any]' to expected argument type 'std.allocator<CInt>'
 30 |
 31 | StdVectorTestSuite.test("VectorOfInt.init(sequence)") {
 32 |     let v = Vector([])
    |                    `- error: cannot convert value of type '[Any]' to expected argument type 'std.allocator<CInt>'
 33 |     expectEqual(v.size(), 0)
 34 |     expectTrue(v.empty())

S:\SourceCache\swift\test\Interop\Cxx\stdlib\use-std-vector.swift:155:15: error: value of type 'Vector' (aka 'std.vector<CInt, allocator<CInt>>') has no member 'map'
153 |     fill(vector: &v)
154 |
155 |     let a = v.map { $0 + 5 }
    |               `- error: value of type 'Vector' (aka 'std.vector<CInt, allocator<CInt>>') has no member 'map'
156 |     expectEqual(a, [6, 7, 8])
157 | }

S:\SourceCache\swift\test\Interop\Cxx\stdlib\use-std-vector.swift:174:14: error: for-in loop requires 'VectorSubclass' to conform to 'Sequence'
172 |
173 |     var count: CInt = 1
174 |     for e in v {
    |              `- error: for-in loop requires 'VectorSubclass' to conform to 'Sequence'
175 |         expectEqual(e, count)
176 |         count += 1

******************** TEST 'Swift(windows-x86_64) :: Interop/CxxToSwiftToCxx/span/span-execution.cpp' FAILED ********************
S:\b\5\tools\swift\test-windows-x86_64\Interop\CxxToSwiftToCxx\span\Output\span-execution.cpp.tmp/use-span.swift:12:15: error: cannot use mutating getter on immutable value: 'sp' is a 'let' constant
10 |
11 | public func printSpan(_ sp: Span) {
12 |   print("{\(sp[0]), \(sp[1]), \(sp[2])}")
   |               `- error: cannot use mutating getter on immutable value: 'sp' is a 'let' constant
13 | }
14 |

S:\b\5\tools\swift\test-windows-x86_64\Interop\CxxToSwiftToCxx\span\Output\span-execution.cpp.tmp/use-span.swift:28:19: error: value of type 'Span' (aka 'std.span<CInt, _CUnsignedLongLong_-1>') has no member 'map'
26 |
27 | public func mapSpan(_ sp: Span) {
28 |   let result = sp.map { $0 + 3 }
   |                   `- error: value of type 'Span' (aka 'std.span<CInt, _CUnsignedLongLong_-1>') has no member 'map'
29 |   print(result)
30 | }
```

--- a/lib/ClangImporter/ImportType.cpp
+++ b/lib/ClangImporter/ImportType.cpp
@@ -2203,10 +2203,11 @@ ImportedType ClangImporter::Implementation::importFunctionReturnType(
   // C++ operators +=, -=, *=, /= may return a reference to self. This is not
   // idiomatic in Swift, let's drop these return values.
   clang::OverloadedOperatorKind op = clangDecl->getOverloadedOperator();
-  if (op == clang::OverloadedOperatorKind::OO_PlusEqual ||
-      op == clang::OverloadedOperatorKind::OO_MinusEqual ||
-      op == clang::OverloadedOperatorKind::OO_StarEqual ||
-      op == clang::OverloadedOperatorKind::OO_SlashEqual)
+  if ((op == clang::OverloadedOperatorKind::OO_PlusEqual ||
+       op == clang::OverloadedOperatorKind::OO_MinusEqual ||
+       op == clang::OverloadedOperatorKind::OO_StarEqual ||
+       op == clang::OverloadedOperatorKind::OO_SlashEqual) &&
+      clangDecl->getReturnType()->isReferenceType())
     return {SwiftContext.getVoidType(), false};
 
   // Fix up optionality.

--- a/lib/IRGen/GenCall.cpp
+++ b/lib/IRGen/GenCall.cpp
@@ -1471,9 +1471,15 @@ void SignatureExpansion::expandExternalSignatureTypes() {
   // Generate function info for this signature.
   auto extInfo = clang::FunctionType::ExtInfo();
 
-  auto &FI = clang::CodeGen::arrangeFreeFunctionCall(IGM.ClangCodeGen->CGM(),
-                                             clangResultTy, paramTys, extInfo,
-                                             clang::CodeGen::RequiredArgs::All);
+  bool isCXXMethod =
+      FnType->getRepresentation() == SILFunctionTypeRepresentation::CXXMethod;
+  auto &FI = isCXXMethod ?
+      clang::CodeGen::arrangeCXXMethodCall(IGM.ClangCodeGen->CGM(),
+          clangResultTy, paramTys, extInfo, {},
+          clang::CodeGen::RequiredArgs::All) :
+      clang::CodeGen::arrangeFreeFunctionCall(IGM.ClangCodeGen->CGM(),
+          clangResultTy, paramTys, extInfo, {},
+          clang::CodeGen::RequiredArgs::All);
   ForeignInfo.ClangInfo = &FI;
 
   assert(FI.arg_size() == paramTys.size() &&
@@ -1595,9 +1601,7 @@ void SignatureExpansion::expandExternalSignatureTypes() {
   if (returnInfo.isIndirect()) {
     auto resultType = getSILFuncConventions().getSingleSILResultType(
         IGM.getMaximalTypeExpansionContext());
-    if (IGM.Triple.isWindowsMSVCEnvironment() &&
-        FnType->getRepresentation() ==
-            SILFunctionTypeRepresentation::CXXMethod) {
+    if (returnInfo.isSRetAfterThis()) {
       // Windows ABI places `this` before the
       // returned indirect values.
       emitArg(0);
@@ -2577,11 +2581,12 @@ public:
 
         // Windows ABI places `this` before the
         // returned indirect values.
-        bool isThisFirst = IGF.IGM.Triple.isWindowsMSVCEnvironment();
-        if (!isThisFirst)
+        auto &returnInfo =
+            getCallee().getForeignInfo().ClangInfo->getReturnInfo();
+        if (returnInfo.isIndirect() && !returnInfo.isSRetAfterThis())
           passIndirectResults();
         adjusted.add(arg);
-        if (isThisFirst)
+        if (returnInfo.isIndirect() && returnInfo.isSRetAfterThis())
           passIndirectResults();
       }
 
@@ -3171,9 +3176,10 @@ void CallEmission::emitToUnmappedMemory(Address result) {
   assert(LastArgWritten == 1 && "emitting unnaturally to indirect result");
 
   Args[0] = result.getAddress();
-  if (IGF.IGM.Triple.isWindowsMSVCEnvironment() &&
-      getCallee().getRepresentation() ==
-          SILFunctionTypeRepresentation::CXXMethod &&
+
+  auto *FI = getCallee().getForeignInfo().ClangInfo;
+  if (FI && FI->getReturnInfo().isIndirect() &&
+      FI->getReturnInfo().isSRetAfterThis() &&
       Args[1] == getCallee().getCXXMethodSelf()) {
     // C++ methods in MSVC ABI pass `this` before the
     // indirectly returned value.
@@ -3566,10 +3572,10 @@ void CallEmission::emitToExplosion(Explosion &out, bool isOutlined) {
       emitToMemory(temp, substResultTI, isOutlined);
       return;
     }
-    if (IGF.IGM.Triple.isWindowsMSVCEnvironment() &&
-        getCallee().getRepresentation() ==
-            SILFunctionTypeRepresentation::CXXMethod &&
-        substResultType.isVoid()) {
+
+    auto *FI = getCallee().getForeignInfo().ClangInfo;
+    if (FI && FI->getReturnInfo().isIndirect() &&
+        FI->getReturnInfo().isSRetAfterThis() && substResultType.isVoid()) {
       // Some C++ methods return a value but are imported as
       // returning `Void` (e.g. `operator +=`). In this case
       // we should allocate the correct temp indirect return

--- a/test/Interop/Cxx/class/method/Inputs/inreg-sret.h
+++ b/test/Interop/Cxx/class/method/Inputs/inreg-sret.h
@@ -1,0 +1,11 @@
+#ifndef TEST_INTEROP_CXX_CLASS_METHOD_INREG_SRET_H
+#define TEST_INTEROP_CXX_CLASS_METHOD_INREG_SRET_H
+
+struct OptionalBridgedBasicBlock {
+};
+
+struct BridgedFunction {
+  OptionalBridgedBasicBlock getFirstBlock() const { return {}; }
+};
+
+#endif // TEST_INTEROP_CXX_CLASS_METHOD_INREG_SRET_H

--- a/test/Interop/Cxx/class/method/Inputs/module.modulemap
+++ b/test/Interop/Cxx/class/method/Inputs/module.modulemap
@@ -19,3 +19,8 @@ module SRetWinARM64 {
   header "sret-win-arm64.h"
   requires cplusplus
 }
+
+module InRegSRet {
+  header "inreg-sret.h"
+  requires cplusplus
+}

--- a/test/Interop/Cxx/class/method/inreg-sret.swift
+++ b/test/Interop/Cxx/class/method/inreg-sret.swift
@@ -1,0 +1,28 @@
+// RUN: %target-swift-emit-irgen -I %S/Inputs -cxx-interoperability-mode=default %s | %FileCheck %s -check-prefix=CHECK -check-prefix=CHECK-%target-cpu
+
+// REQUIRES: OS=windows-msvc
+
+import InRegSRet
+
+final public class BasicBlock {
+}
+
+extension OptionalBridgedBasicBlock {
+  public var block: BasicBlock? { nil }
+}
+
+final public class Function {
+  public var bridged: BridgedFunction {
+    BridgedFunction()
+  }
+
+  public var firstBlock : BasicBlock? { bridged.getFirstBlock().block }
+}
+
+// Check that inreg on the sret isn't missing
+
+// CHECK-x86_64: call void @"?getFirstBlock@BridgedFunction@@QEBA?AUOptionalBridgedBasicBlock@@XZ"(ptr {{.*}}, ptr noalias nocapture sret(%TSo25OptionalBridgedBasicBlockV) {{.*}})
+// CHECK-aarch64: call void @"?getFirstBlock@BridgedFunction@@QEBA?AUOptionalBridgedBasicBlock@@XZ"(ptr {{.*}}, ptr inreg noalias nocapture sret(%TSo25OptionalBridgedBasicBlockV) {{.*}})
+
+// CHECK-x86_64: define {{.*}} void @"?getFirstBlock@BridgedFunction@@QEBA?AUOptionalBridgedBasicBlock@@XZ"(ptr {{.*}} %{{.*}}, ptr noalias sret(%struct.OptionalBridgedBasicBlock) {{.*}} %{{.*}})
+// CHECK-aarch64: define {{.*}} void @"?getFirstBlock@BridgedFunction@@QEBA?AUOptionalBridgedBasicBlock@@XZ"(ptr {{.*}} %{{.*}}, ptr inreg noalias sret(%struct.OptionalBridgedBasicBlock) {{.*}} %{{.*}})

--- a/test/Interop/Cxx/class/method/msvc-abi-return-indirect-trivial-record.swift
+++ b/test/Interop/Cxx/class/method/msvc-abi-return-indirect-trivial-record.swift
@@ -57,7 +57,7 @@ public func test(_ result: VecStr) -> CInt {
 }
 
 // CHECK: swiftcc i32 @"$s4Test4testys5Int32VSo0014vecstr_yuJCataVF"({{.*}} %[[RESULT:.*]])
-// CHECK: call void @"?begin@?$vec@Vstr@@@@QEBA?AVIt@@XZ"(ptr %[[RESULT]], ptr sret{{.*}}
+// CHECK: call void @"?begin@?$vec@Vstr@@@@QEBA?AVIt@@XZ"(ptr %[[RESULT]], ptr {{.*}} sret{{.*}}
 
 public func passTempForIndirectRetToVoidCall() {
     var lhs = LoadableIntWrapper(value: 2)
@@ -65,7 +65,7 @@ public func passTempForIndirectRetToVoidCall() {
     lhs += rhs
 }
 
-// CHECK: void @"$sSo18LoadableIntWrapperV2peoiyyABz_ABtFZ"(ptr
-// CHECK: %[[OPRESULT:.*]] = alloca %struct.LoadableIntWrapper, align 16
-// CHECK-x86_64: call void @"??YLoadableIntWrapper@@QEAA?AU0@U0@@Z"(ptr {{.*}}, ptr sret(%struct.LoadableIntWrapper) %[[OPRESULT]], i32
-// CHECK-aarch64: call void @"??YLoadableIntWrapper@@QEAA?AU0@U0@@Z"(ptr {{.*}}, ptr sret(%struct.LoadableIntWrapper) %[[OPRESULT]], i64
+// CHECK: i32 @"$sSo18LoadableIntWrapperV2peoiyA2Bz_ABtFZ"(ptr
+// CHECK: %[[OPRESULT:.*]] = alloca %TSo18LoadableIntWrapperV, align 4
+// CHECK-x86_64: call void @"??YLoadableIntWrapper@@QEAA?AU0@U0@@Z"(ptr {{.*}}, ptr {{.*}} sret(%TSo18LoadableIntWrapperV) %[[OPRESULT]], i32
+// CHECK-aarch64: call void @"??YLoadableIntWrapper@@QEAA?AU0@U0@@Z"(ptr {{.*}}, ptr {{.*}} sret(%TSo18LoadableIntWrapperV) %[[OPRESULT]], i64

--- a/test/Interop/Cxx/operators/member-inline-irgen.swift
+++ b/test/Interop/Cxx/operators/member-inline-irgen.swift
@@ -5,7 +5,7 @@ import MemberInline
 public func sub(_ lhs: inout LoadableIntWrapper, _ rhs: LoadableIntWrapper) -> LoadableIntWrapper { lhs - rhs }
 
 // CHECK-SYSV: call [[RESA:i32|i64]] [[NAMEA:@_ZN18LoadableIntWrappermiES_]](ptr {{%[0-9]+}}, {{i32|\[1 x i32\]|i64|%struct.LoadableIntWrapper\* byval\(.*\) align 4}} {{%[0-9]+}})
-// CHECK-WIN: call [[RESA:void]] [[NAMEA:@"\?\?GLoadableIntWrapper@@QEAA\?AU0@U0@@Z"]](ptr {{%[0-9]+}}, ptr sret(%struct.LoadableIntWrapper) {{.*}}, i32 {{%[0-9]+}})
+// CHECK-WIN: call [[RESA:void]] [[NAMEA:@"\?\?GLoadableIntWrapper@@QEAA\?AU0@U0@@Z"]](ptr {{%[0-9]+}}, ptr {{.*}} sret(%TSo18LoadableIntWrapperV) {{.*}}, i32 {{%[0-9]+}})
 
 public func call(_ wrapper: inout LoadableIntWrapper, _ arg: Int32) -> Int32 { wrapper(arg) }
 

--- a/test/Interop/Cxx/operators/member-out-of-line-irgen.swift
+++ b/test/Interop/Cxx/operators/member-out-of-line-irgen.swift
@@ -7,5 +7,5 @@ public func add(_ lhs: inout LoadableIntWrapper, _ rhs: LoadableIntWrapper) -> L
 // CHECK-SYSV: call {{i32|i64}} [[NAME:@_ZNK18LoadableIntWrapperplES_]](ptr %{{[0-9]+}}, {{i32|\[1 x i32\]|i64|%struct.LoadableIntWrapper\* byval\(.*\)}}{{.*}})
 // CHECK-SYSV: declare {{.*}}{{i32|i64}} [[NAME]](ptr {{.*}}, {{i32|\[1 x i32\]|i64|%struct.LoadableIntWrapper\* .*byval\(%struct.LoadableIntWrapper\)}}{{.*}})
 
-// CHECK-WIN: call void [[NAME:@"\?\?HLoadableIntWrapper@@QEBA\?AU0@U0@@Z"]](ptr %{{[0-9]+}}, ptr sret(%struct.LoadableIntWrapper) {{.*}}, i32 %{{[0-9]+}})
+// CHECK-WIN: call void [[NAME:@"\?\?HLoadableIntWrapper@@QEBA\?AU0@U0@@Z"]](ptr %{{[0-9]+}}, ptr {{.*}} sret(%TSo18LoadableIntWrapperV) {{.*}}, i32 %{{[0-9]+}})
 // CHECK-WIN: declare dso_local void [[NAME]](ptr {{.*}}, ptr sret(%struct.LoadableIntWrapper) {{.*}}, i32)

